ufs/session.go

```diff
@@ -0,0 +1,357 @@
+package ufs
+
+import (
+	"context"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/user"
+	"path/filepath"
+	"strconv"
+	"sync"
+	"syscall"
+
+	"github.com/docker/go-p9p"
+)
+
+type session struct {
+	sync.Mutex
+	rootRef *FileRef
+	refs    map[p9p.Fid]*FileRef
+}
+
+func NewSession(ctx context.Context, root string) (p9p.Session, error) {
+	return &session{
+		rootRef: &FileRef{Path: root},
+		refs:    make(map[p9p.Fid]*FileRef),
+	}, nil
+}
+
+func (sess *session) getRef(fid p9p.Fid) (*FileRef, error) {
+	sess.Lock()
+	defer sess.Unlock()
+
+	if fid == 0 {
+		return nil, p9p.ErrUnknownfid
+	}
+
+	ref, found := sess.refs[fid]
+	if !found {
+		return nil, p9p.ErrUnknownfid
+	}
+
+	if err := ref.Stat(); err != nil {
+		return nil, err
+	}
+
+	return ref, nil
+}
+
+func (sess *session) newRef(fid p9p.Fid, path string) (*FileRef, error) {
+	sess.Lock()
+	defer sess.Unlock()
+
+	if fid == 0 {
```

Comment:
0 is a valid fid (the Linux kernel uses it as its first fid for example). This should check against `p9p.NOFID`.